posthog/clickhouse/query_fragment.py

```diff
@@ -0,0 +1,99 @@
+import itertools
+import re
+from dataclasses import dataclass
+from functools import cached_property
+from typing import Any, Dict, List, Mapping, Sequence, Tuple, Union
+
+ParamValueType = Union[str, int, float, None, List]
+QueryFragmentLike = Union["QueryFragment", str]
+
+unique_sequence = itertools.count()
+
+
+@dataclass(frozen=True)
+class Param:
+    value: ParamValueType
+
+
+@dataclass
+class QueryFragment:
+    sql: str
+    params: Dict[str, Param]
+
+    def __init__(
+        self,
+        sql: Union[str, "QueryFragment"],
+        params_and_fragments: Mapping[str, Union["Param", "QueryFragment"]] = {},
+        **kwargs: Union["Param", "QueryFragment"],
+    ):
+        if len(kwargs) > 0 and len(params_and_fragments) > 0:
+            raise ValueError("Cannot pass params as both kwargs and via positioned arguments")
+
+        params_and_fragments = params_and_fragments if len(params_and_fragments) > 0 else kwargs
+        if isinstance(sql, str):
+            self.sql = sql
+            self._params = params_and_fragments
+        elif isinstance(sql, QueryFragment):
+            self.sql = sql.sql
+            self._params = sql.params
+
+            if len(params_and_fragments) > 0:
+                raise ValueError("Cannot pass both a QueryFragment object and params to a new QueryFragment")
+
+        self._update_sql()
+
+    @property
+    def query_params(self):
+        "Returns params in format clickhouse expects (removing Param wrapping)"
+        return {key: prop.value for key, prop in self.params.items()}
+
+    @staticmethod
+    def from_tuple(pair: Tuple[str, Any]) -> "QueryFragment":
+        sql, params = pair
+        return QueryFragment(sql, {key: Param(value) for key, value in params.items()})
+
+    @staticmethod
+    def join(by: str, fragments: Sequence[QueryFragmentLike]) -> "QueryFragment":
```

Comment:
not sure what's the idea here?

Reply:
Replacing `' AND '.join(fragments)` in string-building. See the EE cohort query refactor for some examples of this.